src/injected/visualization/element-highlight.tsx

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { CSSProperties } from 'react';
+import * as React from 'react';
+
+import { NamedSFC } from '../../common/react/named-sfc';
+import { FeatureFlagStoreData } from '../../common/types/store-data/feature-flag-store-data';
+import { ClientUtils } from '../client-utils';
+import { RenderDialog } from '../dialog-renderer';
+import { HtmlElementAxeResults } from '../scanner-utils';
+import { DrawerUtils } from './drawer-utils';
+import { DrawerConfiguration, GetBoundingRect } from './formatter';
+import { HightlightBox, HightlightBoxDeps } from './highlight-box';
+
+export type ElementHighlightDeps = {
+    drawerUtils: DrawerUtils;
+    clientUtils: ClientUtils;
+} & HightlightBoxDeps;
+
+export interface ElementHighlightProps {
+    deps: ElementHighlightDeps;
+    element: Element;
+    elementResult: HtmlElementAxeResults;
+    bodyStyle: CSSStyleDeclaration;
+    docStyle: CSSStyleDeclaration;
+    drawerConfig: DrawerConfiguration;
+    featureFlagStoreData: FeatureFlagStoreData;
+    dialogRender: RenderDialog;
+    getBoundingRect: GetBoundingRect;
+}
+
+export const ElementHighlight = NamedSFC<ElementHighlightProps>('HighlightBox', props => {
+    const { deps, drawerConfig, featureFlagStoreData, dialogRender, element, elementResult, bodyStyle, docStyle, getBoundingRect } = props;
+    const { clientUtils, drawerUtils } = deps;
+
+    if (!drawerConfig.showVisualization) {
+        return null;
+    }
+
+    const currentDom = drawerUtils.getDocumentElement();
+    const elementBoundingClientRect = getBoundingRect(element);
+
+    if (drawerUtils.isOutsideOfDocument(elementBoundingClientRect, currentDom, bodyStyle, docStyle)) {
+        return null;
+    }
+
+    const onClickFailureBoxHandler = () => {
+        dialogRender(elementResult, featureFlagStoreData);
+    };
+
+    const offset = clientUtils.getOffset(element);
+    const styles: CSSProperties = {
+        outlineStyle: drawerConfig.outlineStyle,
+        outlineColor: drawerConfig.borderColor,
+        top: drawerUtils.getContainerTopOffset(offset),
+        left: drawerUtils.getContainerLeftOffset(offset),
+        minWidth: drawerUtils.getContainerWidth(offset, currentDom, elementBoundingClientRect.width, bodyStyle, docStyle),
+        minHeight: drawerUtils.getContainerHeight(offset, currentDom, elementBoundingClientRect.height, bodyStyle, docStyle),
+    };
+
+    return (
+        <div title={drawerConfig.toolTip} className={'insights-highlight-box'} style={styles}>
+            <HightlightBox deps={deps} drawerConfig={drawerConfig} boxConfig={drawerConfig.textBoxConfig} className={'text-label'} />
```

Comment:
I'm a little confused with the 2 `<HighlightBox>` here. From looking at the divs of the visualizations on master I can see automated checks visualization uses `failure-label` class name for the inner div that contains the exclamation mark `"!"` but not the `text-label`, while the headings visualization does the opposite (uses `text-label` but no `failure-label`. Is this right? or am I missing something.

If it is right (we use one or the other but not both at the same time), then it seems better to reflect that in this component by using only one `HighlightBox`.
If it is not right (there are case we want both) then maybe we can explore other approaches like using `children` in this  component and let the user class to decide which and how many `HighlihtBox`'s does it needs.

Thoughts?

Reply:
I would re-visit what `createHighlightElement` is doing in `highlight-box-drawer` currently as this is just reproducing what that is doing but with react components.

> If it is not right (there are case we want both) then maybe we can explore other approaches like using children in this component and let the user class to decide which and how many HighlihtBox's does it needs.

Yes, we would want both.

Regarding responsibility: I am not sure I would want to tackle that change right now. From what I understand, this would be a change to handling this in the formatter or something in that vein, since that is how the highlight boxes are configured (so they would seem a far better fit; they all use the same Drawer as of now). I would much rather target moving towards React drawers that behave/maintain design consistency with how our current drawers behave, such that we can move to React drawers, and then start looking at how to better implement responsibility of behavioral differences.

Not sure if that makes sense but the basic gist is I'd like to not change formatters right now. Formatters are the ones responsible for differentiating between visualization differences **implicitly** through drawer configurations. I do think that ideally we should/could move in that direction but I'd like to do that later as my priority would be to move to using react in drawers with our current code.

Let me know if you want to talk more about this offline. 😄

Reply:
It does make sense to apply this changes on a separated PR